titus-common-server/src/main/java/com/netflix/titus/runtime/endpoint/metadata/spring/SpringCallMetadataInterceptor.java

```diff
@@ -0,0 +1,105 @@
+        String callReason = httpServletRequest.getHeader(CallMetadataHeaders.CALL_REASON_HEADER);
+        boolean debug;
+        if (httpServletRequest.getParameter(DEBUG_QUERY_PARAM) == null) {
+            debug = Boolean.parseBoolean(httpServletRequest.getHeader(CallMetadataHeaders.DEBUG_HEADER));
+        } else {
+            debug = Boolean.parseBoolean(httpServletRequest.getParameter(DEBUG_QUERY_PARAM));
+        }
+
+        String originalCallerId = httpServletRequest.getHeader(CallMetadataHeaders.CALLER_ID_HEADER);
+        Authentication delegate = SecurityContextHolder.getContext().getAuthentication();
+
+        Caller directCaller = getDirectCaller(httpServletRequest, delegate);
+
+        CallMetadata.Builder callMetadataBuilder = CallMetadata.newBuilder()
+                .withCallReason(callReason)
+                .withDebug(debug);
+
+        if (originalCallerId == null) {
+            callMetadataBuilder.withCallers(Collections.singletonList(directCaller));
+        } else {
+            CallerType originalCallerType = CallerType.parseCallerType(
+                    originalCallerId,
+                    httpServletRequest.getHeader(CallMetadataHeaders.CALLER_TYPE_HEADER)
+            );
+            Caller originalCaller = Caller.newBuilder()
+                    .withId(originalCallerId)
+                    .withCallerType(originalCallerType)
+                    .build();
+
+            callMetadataBuilder.withCallers(asList(originalCaller, directCaller));
+        }
+
+        CallMetadataAuthentication authentication = new CallMetadataAuthentication(callMetadataBuilder.build(), delegate);
+        SecurityContextHolder.getContext().setAuthentication(authentication);
+
+        return super.preHandle(httpServletRequest, httpServletResponse, handler);
+    }
+
+    protected Caller getDirectCaller(HttpServletRequest httpServletRequest, Authentication delegate) {
+        // If there is an original caller, we know the direct caller must be an application.
+        CallerType callerType = httpServletRequest.getHeader(CallMetadataHeaders.CALLER_ID_HEADER) == null
+                ? CallerType.Unknown
+                : CallerType.Application;
+
+        return Caller.newBuilder()
+                .withId(delegate.getName())
+                .withCallerType(callerType)
+                .withContext(getContextFromServletRequest(httpServletRequest))
+                .build();
+    }
+
+    protected Map<String, String> getContextFromServletRequest(HttpServletRequest httpServletRequest) {
+        Map<String, String> context = new HashMap<>();
+
+        context.put(CallMetadataHeaders.DIRECT_CALLER_CONTEXT_SERVICE_NAME, httpServletRequest.getRequestURI());
+        context.put(CallMetadataHeaders.DIRECT_CALLER_CONTEXT_SERVICE_METHOD, httpServletRequest.getMethod());
+        context.put(CallMetadataHeaders.DIRECT_CALLER_CONTEXT_TRANSPORT_TYPE, "HTTP");
```

Comment:
could use `httpServletRequest.isSecure()` to switch between `HTTP` and `HTTPS` here.

Reply:
Transport type is HTTP or GRPC. Plain/TLS distinction has separate attribute. This code is copy-pasted from the previous (jersey) implementation. I want to keep it the same for compatibility reason.